llmobs: Otel Ingestion GA (#32977)
DataDog/documentation: content/en/llm_observability/instrumentation/otel_instrumentation.md

---
title: OpenTelemetry Instrumentation
---

## Overview
By using OpenTelemetry's standardized semantic conventions for generative AI operations, you can instrument your LLM applications with any OpenTelemetry-compatible library or framework and visualize the traces in LLM Observability.

LLM Observability supports ingesting OpenTelemetry traces that follow the [OpenTelemetry 1.37+ semantic conventions for generative AI][1]. This allows you to send LLM traces directly from OpenTelemetry-instrumented applications to Datadog without requiring the Datadog LLM Observability SDK or a Datadog Agent.

## Prerequisites

- A [Datadog API key][2]
- An application instrumented with OpenTelemetry that emits traces following the [OpenTelemetry 1.37+ semantic conventions for generative AI][1]

> **Note**: [External evaluations][6] are not currently supported for OpenTelemetry spans.

Reviewer: ```suggestion
**Note**: [External evaluations][6] are not supported for OpenTelemetry spans.
```

> <div class="alert alert-info">[External evaluations][6] in LLM Observability are not currently applied to OpenTelemetry spans. Evaluations are only available for spans generated with the Datadog LLM Observability SDK or submitted directly to the HTTP API intake.</div>

Reviewer: Need to use inline links within `<div>`s

```suggestion
<div class="alert alert-info"><a href="/llm_observability/evaluations/external_evaluations">External evaluations</a> in LLM Observability are not applied to OpenTelemetry spans. Evaluations are only available for spans generated with the Datadog LLM Observability SDK or submitted directly to the HTTP API intake.</div>
```

#### Using strands-agents

If you are using the [`strands-agents` library][5], you need to set an additional environment variable to enable traces that are compliant with OpenTelemetry v1.37+:

```
OTEL_SEMCONV_STABILITY_OPT_IN=gen_ai_latest_experimental
```

This environment variable ensures that `strands-agents` emits traces following the OpenTelemetry v1.37+ semantic conventions for generative AI, which are required by LLM Observability.

### Instrumentation

To generate traces compatible with LLM Observability, do one of the following:

- Use an OpenTelemetry library or instrumentation package that emits spans following the [OpenTelemetry 1.37+ semantic conventions for generative AI][1].
- Create custom OpenTelemetry instrumentation that produces spans with the required `gen_ai.*` attributes, as defined in the semantic conventions.

After your application starts sending data, the traces automatically appear in the [**LLM Observability Traces** page][3]. To search for your traces in the UI, use the `ml_app` attribute, which is automatically set to the value of your OpenTelemetry root span's `service` attribute.

> <div class="alert alert-danger">OpenInference and OpenLLMetry are not currently supported, as they have not been updated to support OpenTelemetry 1.37+ semantic conventions for generative AI.</a></div>

Reviewer 2: ```suggestion
<div class="alert alert-danger">
- OpenInference and OpenLLMetry are not currently supported, as they have not been updated to support OpenTelemetry 1.37+ semantic conventions for generative AI
- There may be a 3-5 minute delay between sending traces and seeing them appear on the LLM Observability Traces page. If you have APM enabled, traces appear immediately in the APM Traces page.
</div>
```

> **Note**: There may be a 3-5 minute delay between sending traces and seeing them appear on the LLM Observability Traces page. If you have APM enabled, traces appear immediately in the APM Traces page.

Reviewer 2: ```suggestion
```
